Fix for #4296

Tzenchor opened the pull request:
.gitignore:
 +Added rule for *.csproj.csdat files created by Visual Studio 2015, not needed with the Visual Studio extension, but useful for Git Desktop
 +Added rule for "rebrace.xml" used by the rebracer extension for Visual Studio (this extension allows the user to have one style rules per solution instead of having to change the options manually when changing projects).
TextureImporter.cs:
 *Bitmap files are allways expanded to 32bits and converted to PixelContent<Color>
 *Files with RGBF format are expanded to RGBAF and converted to PixelContent<Vector4>
 +Added check for files with correct headers but bad content
TextureImporterTests.cs:
 +Added 2 tests for bitmaps with less than 8 bits per pixel (one with 4 and another with 1)
 +Added 2 tests for bitmaps with format R5G5B5 and R5G6B5, that were a special case with the old code
 +Added test for reading files with good headers, but bad content.
 *Modified ImportStandard to receive the expected SurfaceFormat for that file
New Files in Test project:
 +Assets\Textures\LogoOnly_64px-4bits.bmp
 +Assets\Textures\LogoOnly_64px-monochrome.bmp
 +Assets\Textures\Logo555.bmp
 +Assets\Textures\Logo565.bmp
 +Assets\Textures\rgbf.tif
Protobuild definitions updated

KonajuGames commented on MonoGame.Framework.Content.Pipeline/TextureImporter.cs:
```diff
@@ -53,7 +53,7 @@ public class TextureImporter : ContentImporter<TextureContent>
         /// Initializes a new instance of TextureImporter.
         /// </summary>
-        public TextureImporter()
+        public TextureImporter( )
```
Reformatting, especially that which is just a whitespace change, is not wanted in a PR.  It adds noise to a PR and hides the real changes.

KonajuGames commented on MonoGame.Framework.Content.Pipeline/TextureImporter.cs:
```diff
@@ -63,168 +63,46 @@ public TextureImporter()
-            FreeImage.ConvertToRawBits(bytes, fBitmap, pitch, bpp, redMask, greenMask, blueMask, true);
+            // Create the byte array for the data,
+            // In case of modifying the code to generate pixel data with less than 8 bpp do not remove the casts or the size or the size will be wrong
+            var bytes = new byte[(int) (width * height * ((float) bpp / 8))];
```
Rather than do the operation as float and convert back to int (and run the risk of floating point inaccuracy), keep it as integer all the way through, and it's easier to read.
`var bytes = new byte[(width * height * bpp) / 8];`

Tzenchor replied:
A good idea, but can lead to problems if in a future someone wants to compress the color data by reducing the bits per pixel.
If bpp is less than 8 your option returns less bytes than needed when the number of pixels does not fill a complete byte.

KonajuGames commented on MonoGame.Framework.Content.Pipeline/TextureImporter.cs:
```diff
@@ -237,5 +115,58 @@ public override TextureContent Import (string filename, ContentImporterContext c
+        /// Switches the red and blue channels
+        /// </summary>
+        /// <param name="fBitmap">image</param>
+        private static void switchRedAndBlueChannels(FIBITMAP fBitmap)
```
Method names always start with an uppercase character.

KonajuGames commented on Test/ContentPipeline/TextureImporterTests.cs:
```diff
@@ -29,51 +29,85 @@ void ImportStandard(string filename)
 
         [Test]
-        public void ImportGif()
+        public void ImportGif( )
```
More unnecessary reformatting.  Also check the [Coding Guidelines](https://github.com/mono/MonoGame/wiki/Coding-Guidelines) for rules regarding spaces.

Tzenchor commented:
I have updated my branch and corrected the offending function name.
I did not change the other style mistakes to avoid further cluttering, but changed my configuration to avoid future problems.
Also i did not modify the byte array generation because of possible future problems

Tzenchor commented:
I have found a problem related to pull request #4136, [DdsLoader.cs](https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework.Content.Pipeline/DdsLoader.cs) can't process a simple texture within a dds file, it will raise a NullReferenceException at line 307
Should i add a fast fix for that here, or do it in another pull?

KonajuGames commented:
Another pull request.

As for future formats not filling a byte, I would not be concerned about
that. Pixel formats will always fill a byte.

Tzenchor commented:
I have found that my code has some flaws than yours
```c#
var bytes = new byte[(int) (width * height * ((float) bpp / 8))];
```
Does exactly the same than 
```c#
var bytes = new byte[(width * height * bpp) / 8];
```
I have been investigating and found the definitive solution in
[stackoverflow](http://stackoverflow.com/a/503201)
```c#
var bytes = new byte[(width * height * bpp -1) / 8 + 1];
```
I'll make the new commits once visual studio finishes its update

KonajuGames commented:
I'm not convinced about that one. If we stay in the integer domain, there
is no rounding or inaccuracy error to worry about.

Tzenchor commented:
I now it sounds weird but works, and remains in the integer domain (in reality they are cast to long because of bpp being uint), did some tests and works in all cases even where the other functions failed. Just try all of them with this values:
Width = 5
Height = 17
bpp = 4
The first two will return 42, that is 1 byte short, while the last returns correctly 43.

KonajuGames commented:
Ok, I'm good with merging this now.  Thanks.
